src/e32.c

```diff
@@ -160,28 +163,24 @@ e32_set_mode(struct E32 *dev, int mode)
 
   int m0 = mode & 0x01;
   int m1 = mode & 0x02;
+  m1 >>= 1;
 
   ret  = gpio_write(dev->gpio_m0_fd, m0) != 1;
   ret |= gpio_write(dev->gpio_m1_fd, m1) != 1;
 
-  if(!ret)
+  if(ret)
   {
-    dev->prev_mode = dev->mode;
-    dev->mode = mode;
-  }
-  else
     return ret;
+  }
 
   if(dev->verbose)
-    debug_output("new mode %d, prev mode is %d\n", mode, dev->prev_mode);
+    debug_output("new mode %d, prev mode is %d\n", dev->mode, dev->prev_mode);
 
-  if(dev->prev_mode == 3 && dev->mode != 3)
+  if(dev->prev_mode != dev->mode)
   {
-    // TODO
-    usleep(54000);
+    usleep(20000);
```

Comment:
This is the main culprit to the hanging. When switching modes if you send commands to the e32 it will not respond.